lightning/src/chain/chainmonitor.rs

```diff
@@ -162,39 +187,31 @@ impl<ChanSigner: ChannelKeys, C: Deref, T: Deref, F: Deref, L: Deref> ChainMonit
 		Ok(())
 	}
 
-	/// Updates the monitor that watches the channel referred to by the given outpoint.
-	fn update_monitor(&self, outpoint: OutPoint, update: ChannelMonitorUpdate) -> Result<(), MonitorUpdateError> {
-		let mut monitors = self.monitors.lock().unwrap();
-		match monitors.get_mut(&outpoint) {
-			Some(orig_monitor) => {
-				log_trace!(self.logger, "Updating Channel Monitor for channel {}", log_funding_info!(orig_monitor));
-				orig_monitor.update_monitor(update, &self.broadcaster, &self.logger)
-			},
-			None => Err(MonitorUpdateError("No such monitor registered"))
-		}
-	}
-}
-
-impl<ChanSigner: ChannelKeys, C: Deref + Sync + Send, T: Deref + Sync + Send, F: Deref + Sync + Send, L: Deref + Sync + Send> chain::Watch for ChainMonitor<ChanSigner, C, T, F, L>
-	where C::Target: chain::Filter,
-	      T::Target: BroadcasterInterface,
-	      F::Target: FeeEstimator,
-	      L::Target: Logger,
-{
-	type Keys = ChanSigner;
-
-	fn watch_channel(&self, funding_txo: OutPoint, monitor: ChannelMonitor<ChanSigner>) -> Result<(), ChannelMonitorUpdateErr> {
-		match self.add_monitor(funding_txo, monitor) {
-			Ok(_) => Ok(()),
-			Err(_) => Err(ChannelMonitorUpdateErr::PermanentFailure),
-		}
-	}
-
 	fn update_channel(&self, funding_txo: OutPoint, update: ChannelMonitorUpdate) -> Result<(), ChannelMonitorUpdateErr> {
-		match self.update_monitor(funding_txo, update) {
-			Ok(_) => Ok(()),
-			Err(_) => Err(ChannelMonitorUpdateErr::PermanentFailure),
+		// Update the monitor that watches the channel referred to by the given outpoint.
+		let mut monitors = self.monitors.lock().unwrap();
+		if let Some(orig_monitor) = monitors.get_mut(&funding_txo) {
+			log_trace!(self.logger, "Updating Channel Monitor for channel {}", log_funding_info!(orig_monitor));
+			let res = orig_monitor.update_monitor(&update, &self.broadcaster, &self.logger);
+			// Even if updating the monitor returns an error, the monitor's state will
+			// still be changed. So, persist the updated monitor despite the error.
+			match self.persister.update_persisted_channel(funding_txo, &update, orig_monitor) {
+				Err(e) => {
+					if !res.is_err() {
+						return Err(e);
+					}
+				},
+				_ => {}
+			}
+			if let Err(e) = res {
+				log_error!(self.logger, "{:?}", e);
+				return Err(ChannelMonitorUpdateErr::PermanentFailure);
+			}
+		} else {
+			log_error!(self.logger, "Failed to update channel monitor: no such monitor registered");
+			return Err(ChannelMonitorUpdateErr::PermanentFailure);
 		}
```

Comment:
Consider making this a `match` with `None` as the first arm. Then it would come across more as a guard clause and be more easily connected with the source of the error.